src/db.c

```diff
@@ -1070,6 +1070,104 @@ void moveCommand(client *c) {
+    robj *o;
+    redisDb *src, *dst;
+    int srcid;
+    long long dbid, expire;
+    int j, replace = 0, delete = 0;
+    
+    if (c->argc >= 6){
+        addReply(c,shared.syntaxerr);
+        return;
+    }
+    
+    /* Obtain source and target DB pointers 
+     * Default target DB is the same as the source DB 
+     * Parse the REPLACE option and targetDB option. */
+    src = c->db;
+    srcid = c->db->id;
+    dbid = c->db->id;
+    for (j = 3; j < c->argc; j++) {
+        if (!strcasecmp(c->argv[j]->ptr,"replace")) {
+            replace = 1;
+        } else if (getLongLongFromObject(c->argv[j],&dbid) == C_ERR ||
+            dbid < INT_MIN || dbid > INT_MAX ||
+            selectDb(c,dbid) == C_ERR)
+        {
+            addReply(c,shared.outofrangeerr);
+            return;
+        }
+    }
+    dst = c->db;
+    selectDb(c,srcid); /* Back to the source DB */
+    if ((server.cluster_enabled == 1) && (srcid != 0 || dbid != 0)) {
+        addReplyError(c,"Copying to another database is not allowed in cluster mode");
+        return;
+    }
+
+    /* If the user select the same DB as
+     * the source DB and using newkey as the same key
+     * it is probably an error. */
+    robj *key = c->argv[1];
+    robj *newkey = c->argv[2];
+    if (src == dst && (sdscmp(key->ptr, newkey->ptr) == 0)) {
+        addReply(c,shared.sameobjecterr);
+        return;
+    }
+
+    /* Check if the element exists and get a reference */
+    o = lookupKeyWrite(c->db,key);
+    if (!o) {
+        addReply(c,shared.czero);
+        return;
+    }
+    expire = getExpire(c->db,key);
+
+    /* Return zero if the key already exists in the target DB. 
+     * If REPLACE option is selected, delete newkey from targetDB. */
+    if (lookupKeyWrite(dst,newkey) != NULL) {
+        if (replace) {
+            delete = 1;
+        } else {
+            addReply(c,shared.czero);
+            return;
+        }
+    }
+
+    /* Duplicate object according to object's type. */
+    robj *newobj;
+    switch(o->type) {
+        case OBJ_STRING: newobj = dupStringObject(o); break;
+        case OBJ_LIST: newobj = dupListObject(o); break;
+        case OBJ_SET: newobj = dupSetObject(o); break;
+        case OBJ_ZSET: newobj = dupZsetObject(o); break;
+        case OBJ_HASH: newobj = dupHashObject(o); break;
+        case OBJ_STREAM: newobj = dupStreamObject(o); break;
+        case OBJ_MODULE:
+            addReplyError(c, "Copying　module type object is not supported ");
+            return;
+        default: {
+            addReplyError(c, "unknown type object");
+            return;
+        };
+    }
+
+    if (delete) {
+        dbDelete(dst,newkey);
+    }
+
+    dbAdd(dst,newkey,newobj);
+    if (expire != -1) setExpire(c, dst, newkey, expire);
+
+    /* OK! key copied */
+    signalModifiedKey(c,dst,c->argv[2]);
+    notifyKeyspaceEvent(NOTIFY_GENERIC,"copy_to",c->argv[2],dst->id);
+
+    server.dirty++;
+    addReply(c,shared.cone);
```

Comment:
Not sure how I feel about this specifically, but we maybe should retain the LRU information. We could also document the fact we just don't do that.

Reply:
we recently made an effort to replace many of the dbDelete+dbAdd pairs with setKey (see SUNIONSTORE and alike).
The difference is that `setKey` retains the LRU of the destination key.

However, my above note is actually unrelated since we're talking of the LRU of the source key.
from what i can tell, both MOVE and RENAME, retain the LRU of the source key, so i'm guessing COPY should do the same?

Or if we consider that a new object (unlike MOVE and RENAME), we should behave similarly to RESTORE REPLACE (create a fresh new key)?
or maybe this is in some way similar to SUNIONSTORE being called with one input (clones a set and keeps the destination LRU)?

@redis/core-team @guybe7 please share your thoughts. (i'm leaning towards the RESTORE REPLACE behavior)

Reply:
i vote for:
src's LRU should remain unchanged (like MOVE)
dst's LRU should be changed (like RESTORE REPLACE)